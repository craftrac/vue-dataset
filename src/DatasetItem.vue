<template>
  <component :is="tag">
    <template v-for="(rowIndex, i) in dsRows">
      <slot :row="dsData[rowIndex]" :row-index="rowIndex" :index="indexes[i]"></slot>
    </template>
    <slot v-if="!dsRows.length" name="noDataFound"></slot>
  </component>
</template>

<script>
import { computed, inject } from 'vue'

export default {
  name: 'DatasetItem',
  props: {
    tag: {
      type: String,
      default: 'div',
    },
  },
  setup() {
    const indexes = computed(() => {
      const arr = []

      for (let i = inject('dsFrom').value; i < inject('dsTo').value; i++) {
        arr.push(i)
      }

      return arr
    })

    return {
      dsData: inject('dsData'),
      dsRows: inject('dsRows'),
      indexes,
    }
  },
}
</script>
