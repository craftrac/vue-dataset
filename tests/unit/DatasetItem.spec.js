import { mount } from '@vue/test-utils'
import { ref } from 'vue'
import DatasetItem from '@/DatasetItem.vue'

describe('DatasetItem', () => {
  let wrapper = null

  function wrapperWithProvide(provideOpts) {
    const wrapper = mount(DatasetItem, {
      global: {
        provide: {
          dsData: ref([
            {
              age: 20,
              name: 'Jessie Casey',
              email: 'jessie.casey@flyboyz.biz',
            },
            {
              age: 26,
              name: 'Solomon Stanley',
              email: 'solomon.stanley@tetak.net',
            },
          ]),
          dsRows: ref([0, 1]),
          dsFrom: ref(0),
          dsTo: ref(10),
          ...provideOpts,
        },
      },
      slots: {
        default: `
          <template v-slot="{ row, rowIndex, index }">
            <div class="result">
              <p>{{ row.name }}</p>
              <p>{{ rowIndex }}</p>
              <p>{{ index }}</p>
            </div>
          </template>
        `,
        noDataFound: `<p>No results found</p>`,
      },
    })

    return wrapper
  }

  afterEach(function () {
    wrapper.unmount()
  })

  it('renders correctly', () => {
    wrapper = wrapperWithProvide()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders divs based on passed props', () => {
    wrapper = wrapperWithProvide()
    expect(wrapper.findAll('div.result').length).toBe(2)
  })

  it('calculates the correct indexes', () => {
    wrapper = wrapperWithProvide()
    expect(wrapper.vm.indexes).toEqual([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
  })

  it('does not render any results when dsRows is empty', () => {
    wrapper = wrapperWithProvide({
      dsRows: ref([]),
    })
    expect(wrapper.findAll('div.result').length).toBe(0)
  })

  it('renders the noDataFound slot when dsRows is empty', () => {
    wrapper = wrapperWithProvide({
      dsRows: ref([]),
    })
    expect(wrapper.find('p').text()).toBe('No results found')
  })

  it('renders divs after data changed', () => {
    wrapper = wrapperWithProvide({
      dsData: ref([
        {
          age: 17,
          name: 'John Doe',
          email: 'john.doe@flyboyz.biz',
        },
      ]),
      dsRows: ref([0]),
    })
    expect(wrapper.findAll('div.result').length).toBe(1)
    expect(wrapper.element).toMatchSnapshot()
  })
})
