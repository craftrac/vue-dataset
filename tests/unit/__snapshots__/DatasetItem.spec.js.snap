// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DatasetItem > renders correctly 1`] = `
<div>
  
  
  <div
    class="result"
  >
    <p>
      Jessie Casey
    </p>
    <p>
      0
    </p>
    <p>
      0
    </p>
  </div>
  
  
  <div
    class="result"
  >
    <p>
      Solomon Stanley
    </p>
    <p>
      1
    </p>
    <p>
      1
    </p>
  </div>
  
  
  <!--v-if-->
</div>
`;

exports[`DatasetItem > renders divs after data changed 1`] = `
<div>
  
  
  <div
    class="result"
  >
    <p>
      John Doe
    </p>
    <p>
      0
    </p>
    <p>
      0
    </p>
  </div>
  
  
  <!--v-if-->
</div>
`;
